src/main/java/org/jenkinsci/deprecatedusage/Main.java

```diff
@@ -1,75 +1,167 @@
+                        UpdateCenter updateCenter = new UpdateCenter(jsonRoot, dir);
+                        cores.add(updateCenter.getCore());
+                        plugins.addAll(updateCenter.getPlugins());
+                    }
+
+                    @Override
+                    public void failed(Exception ex) {
+                        System.out.println("Error downloading update center metadata for " + updateCenterURL);
+                        System.out.println(ex.toString());
+                        metadataLoaded.countDown();
+                    }
 
-        System.out.println("duration : " + (System.currentTimeMillis() - start) + " ms at "
-                + DateFormat.getDateTimeInstance().format(new Date()));
+                    @Override
+                    public void cancelled() {
+                        System.out.println("Cancelled download of " + updateCenterURL);
+                    }
+                });
+            }
+
+            // wait for async code to finish submitting
+            metadataLoaded.await(15, TimeUnit.SECONDS);
+            System.out.println("Downloading core files");
+            CompletableFuture<Void> coreAnalysisComplete = CompletableFuture.allOf(
+                    cores.stream()
+                            .map(core -> core.downloadIfNotExists(client, threadPool).thenRun(() -> {
+                                try {
+                                    System.out.println("Analyzing deprecated APIs in " + core);
+                                    deprecatedApi.analyze(core.getFile());
+                                    System.out.println("Finished deprecated API analysis in " + core);
+                                } catch (IOException e) {
+                                    System.out.println("Error analyzing deprecated APIs in " + core);
+                                    System.out.println(e.toString());
+                                }
+                            }))
+                            .toArray(CompletableFuture[]::new));
+
+            System.out.println("Downloading plugin files");
+            int pluginCount = plugins.size();
+            int maxConcurrent = Math.max(options.maxConcurrentDownloads, Runtime.getRuntime().availableProcessors() * 4);
+            Semaphore concurrentDownloadsPermit = new Semaphore(maxConcurrent);
+            List<JenkinsFile> downloadedPlugins = Collections.synchronizedList(new ArrayList<>(pluginCount));
+            List<CompletableFuture<?>> futures = new ArrayList<>(pluginCount + 1);
+            futures.add(coreAnalysisComplete);
+            for (JenkinsFile plugin : plugins) {
+                concurrentDownloadsPermit.acquire();
+                futures.add(plugin.downloadIfNotExists(client, threadPool).handle((success, failure) -> {
+                    concurrentDownloadsPermit.release();
+                    if (failure != null) {
+                        System.out.println("Error downloading " + plugin);
```

Comment:
Shouldn't this rather be a URL?

    Error downloading work/59a92455/ibm-application-security/1.3.2/ibm-application-security.hpi

Doesn't seem all that useful.

Reply:
Ha, yeah, the logging in this could all be a little improved it seems. Thanks for the ideas!